myia/opt/rewrite.py

```diff
@@ -1,70 +1,472 @@
+
+        The arguments that are added are the original free variables, or
+        DEAD if none of the graphs that can be called at that site have that
+        free variable.
+        """
+        fvs = [
+            fv
+            if any(fv in gg.free_variables_extended for gg in entry.calls[node])
+            else make_dead(fv)
+            for fv in entry.fvs
+        ]
+        new_node = node.graph.apply(*node.inputs, *fvs)
+        new_node.abstract = node.abstract
+        self.manager.replace(node, new_node)
+        return True
+
+    def rewrite_graph(self, entry):
+        """Rewrite the graphs.
+
+        New parameters are added for each free variable.
+
+        Then, we redirect all free variables within scope to the new
+        parameters, which means that they are not closures anymore.
+        """
+        mng = self.manager
+        new_params = list(entry.graph.parameters)
+        with mng.transact() as tr:
+            # Redirect the fvs to the parameter (those in scope)
+            for fv in entry.fvs:
+                param = self.param(entry.graph, fv)
+                new_params.append(param)
+                if fv in entry.graph.free_variables_extended:
+                    for node, idx in mng.uses[fv]:
+                        if node.graph in entry.scope:
+                            tr.set_edge(node, idx, param)
+            tr.set_parameters(entry.graph, new_params)
+        return True
+
+
+class GraphInterfaceRewriterOpt(Partializable):
+    """Implements optimizer interface for GraphInferfaceRewriter."""
+
+    def __init__(self, resources, rewriter):
+        """Initialize GraphInterfaceRewriterOpt.
+
+        Arguments:
+            resources: The resources object associated to the pipeline.
+            rewriter: A subclass of GraphInterfaceRewriter. It will be
+                instantiated in the __call__ method.
+        """
         self.resources = resources
-        self.name = "rewrite_graphs"
+        self.rewriter = rewriter
+        self.name = rewriter.__name__
 
     def __call__(self, root):
-        """Apply rewrite_graphs on root."""
-        args = dict(
-            opt=self,
-            node=None,
-            manager=self.resources.opt_manager,
-            profile=False,
-        )
+        """Apply the rewriter on root."""
+        mng = self.resources.opt_manager
+        args = dict(opt=self, node=None, manager=mng, profile=False,)
         with tracer("opt", **args) as tr:
             tr.set_results(success=False, **args)
-            chg = rewrite_graphs(root, self.resources.opt_manager)
+            mng.gc()
+            rewriter = self.rewriter(mng)
+            chg = rewriter.run()
             if chg:
                 tracer().emit_success(**args, new_node=None)
             return chg
 
 
-__all__ = ["RewriteGraphs", "rewrite_graphs"]
+##########################
+# LAMBDA LIFTING EXAMPLE #
+##########################
```

Comment:
This should probably go with the LambdaLiftRewriter above

Reply:
OK